lunatrace/bsl/backend/src/snapshot/package-tree.ts

```diff
@@ -0,0 +1,165 @@
+    // pop value from stack
+    const next = stack.pop();
+    if (!next) {
+      break;
+    }
+
+    stack.push(...depDependencies(next.dependency, next.id));
+    res.push(next);
+  }
+  // reverse to restore input order
+  return res;
+}
+
+function mapPackageTreeToBuildDependencyRelationships(
+  buildId: string,
+  pkgTree: PkgTree
+): Build_Dependency_Relationship_Insert_Input[] {
+  const flattenedDependencies = flattenPkgTree(pkgTree);
+  return flattenedDependencies
+    .map((entry): Build_Dependency_Relationship_Insert_Input | null => {
+      const { id, dependency, parentId } = entry;
+      if (!dependency.version || !dependency.name) {
+        return null;
+      }
+
+      return {
+        /*
+         TODO (cthompson) we are generating the ids for the database here. This avoids a roundtrip from the server
+         so we can have an id on hand for `depended_by_relationship_id` but could result in an insertion error.
+         What is the best way to recover from this? Theoretically, on error, we can
+         regenerate the graphql insertion query which would create new UUIDs which would be incredibly unlikely to have another
+         collision. For now, this is most likely fine and shouldn't result in any issues, the planets need to align for this to
+         be a problem atm.
+         */
+        id,
+        build_id: buildId,
+        depended_by_relationship_id: parentId,
+        release: {
+          data: {
+            version: dependency.version,
+            package: {
+              data: {
+                name: dependency.name,
+                package_manager: 'npm',
+                // TODO (cthompson) theoretically we have the information to fill this in, how do we get this from a PkgTree?
+                custom_registry: '',
+              },
+              on_conflict: {
+                constraint: Package_Constraint.PackagePackageManagerCustomRegistryNameIdx,
+                update_columns: [
+                  Package_Update_Column.Name,
+                  Package_Update_Column.PackageManager,
+                  Package_Update_Column.CustomRegistry,
+                ],
+              },
+            },
+          },
+          on_conflict: {
+            constraint: Package_Release_Constraint.ReleasePackageIdVersionIdx,
+            update_columns: [Package_Release_Update_Column.Version],
+          },
+        },
+        labels: dependency.labels,
+        range: dependency.range,
+      };
+    })
+    .filter(notEmpty);
+}
+
+export async function snapshotPinnedDependencies(buildId: string, repoDir: string) {
+  const pkgTrees = await collectPackageTreesFromDirectory(repoDir);
+
+  const buildDependencyRelationships = pkgTrees
+    .map((pkgTree) => mapPackageTreeToBuildDependencyRelationships(buildId, pkgTree))
+    .flat();
+
+  log.info('generated package tree for pinned dependencies', {
+    count: buildDependencyRelationships.length,
+  });
+
+  const chunkSize = 5000;
```

Comment:
Do we actually need chunking here?  How can we be sure it wont insert a deeper item in the tree before it inserts a higher item and violate the foreign key constraint? Are they in order? Can we just try doing them flat until it breaks? Are we hitting some built in limit in hasura?  They have some security stuff around this in their enterprise stuff but its really hard to figure out whats going on in our OSS version. https://hasura.io/docs/latest/graphql/core/databases/postgres/mutations/upsert/#pg-nested-upsert-caveats